# Dictionary

## iteritems -> items and faster due to [view](https://docs.python.org/3.3/library/stdtypes.html#dict-views), similar to iterator


In [175]:
!python -V && exit

Python 3.10.5


range is a generator in python3

In [176]:
range(5)

range(0, 5)

In [ ]:
# from itertools import zip_longest
a = zip(range(1000), range(0, 1000, 2))
dictionary = dict(a)
dictionary

In [178]:
count = 0
for k,v in dictionary.items():
    if k%2 and v%2: count+=1

Now we can reverse a dictionary view,

In [179]:
list(reversed(dictionary))

[499,
 498,
 497,
 496,
 495,
 494,
 493,
 492,
 491,
 490,
 489,
 488,
 487,
 486,
 485,
 484,
 483,
 482,
 481,
 480,
 479,
 478,
 477,
 476,
 475,
 474,
 473,
 472,
 471,
 470,
 469,
 468,
 467,
 466,
 465,
 464,
 463,
 462,
 461,
 460,
 459,
 458,
 457,
 456,
 455,
 454,
 453,
 452,
 451,
 450,
 449,
 448,
 447,
 446,
 445,
 444,
 443,
 442,
 441,
 440,
 439,
 438,
 437,
 436,
 435,
 434,
 433,
 432,
 431,
 430,
 429,
 428,
 427,
 426,
 425,
 424,
 423,
 422,
 421,
 420,
 419,
 418,
 417,
 416,
 415,
 414,
 413,
 412,
 411,
 410,
 409,
 408,
 407,
 406,
 405,
 404,
 403,
 402,
 401,
 400,
 399,
 398,
 397,
 396,
 395,
 394,
 393,
 392,
 391,
 390,
 389,
 388,
 387,
 386,
 385,
 384,
 383,
 382,
 381,
 380,
 379,
 378,
 377,
 376,
 375,
 374,
 373,
 372,
 371,
 370,
 369,
 368,
 367,
 366,
 365,
 364,
 363,
 362,
 361,
 360,
 359,
 358,
 357,
 356,
 355,
 354,
 353,
 352,
 351,
 350,
 349,
 348,
 347,
 346,
 345,
 344,
 343,
 342,
 341,
 340,
 339,
 338,
 337,
 336,
 335,
 334,
 333

Get back a readonly dictionary view? Available only in python 3.10

In [180]:
values = dictionary.values()
values.mapping

mappingproxy({0: 0,
              1: 2,
              2: 4,
              3: 6,
              4: 8,
              5: 10,
              6: 12,
              7: 14,
              8: 16,
              9: 18,
              10: 20,
              11: 22,
              12: 24,
              13: 26,
              14: 28,
              15: 30,
              16: 32,
              17: 34,
              18: 36,
              19: 38,
              20: 40,
              21: 42,
              22: 44,
              23: 46,
              24: 48,
              25: 50,
              26: 52,
              27: 54,
              28: 56,
              29: 58,
              30: 60,
              31: 62,
              32: 64,
              33: 66,
              34: 68,
              35: 70,
              36: 72,
              37: 74,
              38: 76,
              39: 78,
              40: 80,
              41: 82,
              42: 84,
              43: 86,
              44: 88,
              45: 90,
   

In [181]:
values.mapping[228]

456

update dictionary

In [182]:
dictionary_1 = {'a': 1, 'b': 2}
dictionary_2 = {'b': 3, 'c': 4}
%time dictionary_1 |= dictionary_2
print(dictionary_1)

CPU times: user 1 µs, sys: 1 µs, total: 2 µs
Wall time: 3.1 µs
{'a': 1, 'b': 3, 'c': 4}


merge dictionary

In [183]:
merge_dictionary_1 = {**dictionary_1, **dictionary_2}
print(merge_dictionary_1)

{'a': 1, 'b': 3, 'c': 4}


In [184]:
dictionary_1 = {'a': 1, 'b': 2}
%time dictionary_1 =  {'b': 3, 'c': 4} | dictionary_1
dictionary_1
# show screenshot

CPU times: user 2 µs, sys: 0 ns, total: 2 µs
Wall time: 3.1 µs


{'b': 2, 'c': 4, 'a': 1}

Assignment operator
so many times we have to define a variable and use that variable under a condition

In [185]:
# Python2
x = 5
if x:
    print(x)

5


In [186]:
# python 3
if x:=5:
    print(x)

5


Merge a list

In [187]:
lst = ['d', 'e', 'f']
['a','b','c', *lst]

['a', 'b', 'c', 'd', 'e', 'f']

Ignore _ in numerals

In [188]:
123_667.0_1

123667.01

What will be the output?

In [189]:
round(16.5, 10)

16.5

Global namespace

In [190]:

[x for x in [1,2,5]]
x

5

In [191]:
3/2

1.5

In [192]:
3//2

1

In [193]:
bigone = 12345678901234567890
print(type(bigone))

<class 'int'>


In [194]:
print("3" > 4)

TypeError: '>' not supported between instances of 'str' and 'int'

In [195]:
a = zip(range(100), range(100))
a

In [196]:
# a.next() python 2
next(a)

(0, 0)

In [198]:
import string
pairs = zip(string.ascii_uppercase, string.ascii_lowercase)
pairs

In [199]:
for i in b"hi":
    print(i, type(i))

104 <class 'int'>
105 <class 'int'>


In [200]:
print("café" + u"hello")

caféhello


Strings are in Unicode by default in Python 3.x

In [201]:
russian_hellé = 'Здравствуйте'
erh = russian_hellé.encode('utf-8')
erh
#type(erh)

b'\xd0\x97\xd0\xb4\xd1\x80\xd0\xb0\xd0\xb2\xd1\x81\xd1\x82\xd0\xb2\xd1\x83\xd0\xb9\xd1\x82\xd0\xb5'

In [202]:
class ABC:
    def __init__(self, x, y):
        super().__init__(x, y)
    

In [203]:
x=5.123
y=2
print(f'my fav number {x:.{y}f}')

my fav number 5.12


print out the variable and the value of the variable

In [204]:
print(f'{x=}')

x=5.123


In [205]:
2 > (1,2)

TypeError: '>' not supported between instances of 'int' and 'tuple'

In [206]:
b = `12` # no longer works in python 3
b

SyntaxError: invalid syntax (271657561.py, line 1)

## some credits to https://devopedia.org/python-2-vs-3